common/app/conf/HealthCheck.scala

```diff
@@ -0,0 +1,153 @@
+    new HealthCheckResult(url, Right(status))
+  }
+  def apply(url: String, error: Throwable) = {
+    new HealthCheckResult(url, Left(error))
+  }
+  def apply(url: String, status: Int, date: DateTime) = {
+    new HealthCheckResult(url, Right(status), date)
+  }
+  def apply(url: String, error: Throwable, date: DateTime) = {
+    new HealthCheckResult(url, Left(error), date)
+  }
+  def apply(url: String, status: Int, date: DateTime, expiration: Duration) = {
+    new HealthCheckResult(url, Right(status), date, expiration)
+  }
+  def apply(url: String, error: Throwable, date: DateTime, expiration: Duration) = {
+    new HealthCheckResult(url, Left(error), date, expiration)
+  }
+}
+
+private[conf] trait HealthCheckFetcher extends ExecutionContexts with Logging {
+
+  import play.api.Play.current
+
+  def testPort: Int
+
+  lazy val port = {
+    Play.current.mode match {
+      case Mode.Test => testPort
+      case _ => 9000
+    }
+  }
+
+  lazy val baseUrl = s"http://localhost:$port"
+
+  protected def fetchResult(path: String): Future[HealthCheckResult] = {
+    WS.url(s"$baseUrl$path")
+      .withHeaders("X-Gu-Management-Healthcheck" -> "true")
+      .withRequestTimeout(10000).get()
+      .map(result => HealthCheckResult(path, Right(result.status)))
+      .recover {
+        case NonFatal(t) =>
+          log.error(s"HealthCheck request to ${path} failed", t)
+          HealthCheckResult(path, Left(t))
+      }
+  }
+
+  protected def fetchResults(paths: String*): Seq[Future[HealthCheckResult]] = paths.map(fetchResult)
+
+}
+
+private[conf] trait HealthCheckCache extends HealthCheckFetcher {
+
+  protected val cache = AkkaAgent[Map[String, HealthCheckResult]](Map.empty)
+  def get() = cache.get()
+
+  def allSuccessful = cache.get().values.toList match {
+    case Nil => false
+    case nonEmpty => nonEmpty.forall(_.recentlySucceed)
+  }
+  def anySuccessful = cache.get().values.exists(_.recentlySucceed)
+
+  def fetchPaths(paths: Seq[String]) = {
+    log.info("Fetching HealthChecks...")
+    val healthCheckResults = fetchResults(paths:_*)
+
+    Future.sequence(healthCheckResults)
+      .map { m =>
+        m.foreach { case healthCheckResult =>
+          cache.alter(_.updated(healthCheckResult.url, healthCheckResult))
+        }
+      }
+  }
+
+}
+
+trait CachedHealthCheckController extends Controller with Results with ExecutionContexts with Logging {
+
+  val paths: Seq[String]
+  val port: Int
+
+  def healthCheck(): Action[AnyContent]
+  private val cache = new HealthCheckCache {
+    override def testPort: Int = port
+  }
+
+  def runChecks = cache.fetchPaths(paths)
+
+  private def healthCheckResponse(condition: => Boolean) = Action.async {
+    Future.successful {
+      val response = cache.get().map {
+        case (url, r) => s"GET ${url} '${r.formattedResult}' '${r.formattedDate}'"
+      }
+        .mkString("\n")
+      if(condition) Ok(response) else ServiceUnavailable(response)
```

Comment:
We are currently returning an `Internal Server error` if /_healthcheck fails. https://github.com/guardian/frontend/blob/master/common/app/conf/AllGoodHealthcheckController.scala#L59

I think it make more sense to return `Service unavailable`
What do you think?